### PR TITLE
Support deterministic evaluation for onpolicy algorithms

### LINKED ISSUE #353: Deterministic sampling for PGPolicy
- [ ] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [ ] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [x] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [x] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, numpy, sys
  print(tianshou.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  ```

```
0.4.1 1.8.0 1.19.2 3.8.8 (default, Feb 24 2021, 21:46:12) 
[GCC 7.3.0] linux
```

I was wondering why tianshou's PGPolicy, A2CPolicy and PPOPolicy always use `dist.sample()`. This causes undeterminstic behavior. On the other hand, many other RL frameworks (e.g., rllib) provides ways to make it determinstic by using argmax (https://github.com/ray-project/ray/blob/cecfc3b43b7a1a7ec4398645556ef64c21704ea8/rllib/models/torch/torch_action_dist.py#L362).

It would be nice if tianshou can provide an init parameter that enables deterministic sampling in case of evaluation mode, e.g.,

```python
if self.training:
  act = self.dist_fn(logits).sample()
else:
  act = logits.argmax()
```

or maybe can we make the `dist_fn` aware of current mode?

### PR DESCRIPTION
**Pull request by ultmaster**
- [x] I have marked all applicable categories:
    + [ ] exception-raising fix
    + [ ] algorithm implementation fix
    + [ ] documentation modification
    + [x] new feature
- [x] If applicable, I have mentioned the relevant/related issue(s) #353

Less important but also useful:

- [x] I have visited the [source website](https://github.com/thu-ml/tianshou)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [x] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, numpy, sys
  print(tianshou.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  ```

### CODE REVIEW

File: tianshou/policy/modelfree/pg.py [hunk 1]
```diff
@@ -103,7 +108,13 @@ def forward(
             dist = self.dist_fn(*logits)
         else:
             dist = self.dist_fn(logits)
-        act = dist.sample()
+        if self._deterministic_eval and not self.training:
+            if isinstance(logits, tuple):
+                act = logits[0].argmax(-1)
```
Comment (Trinkle23897) [inline tianshou/policy/modelfree/pg.py]: Currently tuple means `(mu, sigma)` for continuous action space, and it seems like using `act = logits[0]` is better.

Another thing: we should check the action space's type in this case, discrete or continuous.

Comment (ultmaster) [reply tianshou/policy/modelfree/pg.py]: Maybe we can check whether dist_fn is `torch.distributions.Categorical`. If it is, we apply an extra argmax?

Does it make sense?

I think it would be more convenient if `torch.distributions.Distribution` had a `deterministic_sample()`, but it didn't...

Comment (Trinkle23897) [reply tianshou/policy/modelfree/pg.py]: > Maybe we can check whether dist_fn is `torch.distributions.Categorical`.

That's not a universal approach. If someone wants to use another dist_fn for discrete action space and `deterministic_eval=True`, it won't result in the expected behavior.

Comment (ultmaster) [reply tianshou/policy/modelfree/pg.py]: Yes. But without argmax, the action shape (action space) would be different in training and evaluation. I guess it's not a good idea to change the corresponding environment in evaluation. What's your suggestion?

Can we enforce user to implement a `determinstic_sample` in `dist_fn` if they want to use this feature, and throw an error if flag is true but the method is not implemented?

Comment (Trinkle23897) [reply tianshou/policy/modelfree/pg.py]: > Can we enforce user to implement

Better not. It would be much difficult when using DiagGaussian distribution.

> What's your suggestion?

Well, my current thought is to check the type of action_space in `BasePolicy.__init__`, and use it here.

Comment (ultmaster) [reply tianshou/policy/modelfree/pg.py]: Make sense to me. Implementation updated.

File: tianshou/policy/base.py [hunk 2]
```diff
@@ -66,6 +67,11 @@ def __init__(
         super().__init__()
         self.observation_space = observation_space
         self.action_space = action_space
+        self.action_type = ""
+        if isinstance(action_space, (Discrete, MultiDiscrete, MultiBinary)):
+            self.action_type = "discrete"
+        elif isinstance(action_space, Box):
+            self.action_type = "continuous"
```
Comment (ultmaster) [inline tianshou/policy/base.py:74]: I think `gym.spaces` is more than 4 types. Would `self.action_type == ""` cause any problems?

Comment (Trinkle23897) [reply tianshou/policy/base.py:74]: "" eventually handles Dict/Tuple ~or other types~ no other types.

**Review by danagi — APPROVED** [review]

**Review by ChenDRAG — APPROVED** [review]
